### PR TITLE
[clang][HLSL][SPRI-V] Add convergence intrinsics

### PR DESCRIPTION
**Pull request by Keenuts**
HLSL has wave operations and other kind of function which required the control flow to either be converged, or respect certain constraints as where and how to re-converge.

At the HLSL level, the convergence are mostly obvious: the control flow is expected to re-converge at the end of a scope.
Once translated to IR, HLSL scopes disapear. This means we need a way to communicate convergence restrictions down to the backend.

For this, the SPIR-V backend uses convergence intrinsics. So this commit adds some code to generate convergence intrinsics when required.

### CODE REVIEW

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (arsenm) [inline clang/lib/CodeGen/CGBuiltin.cpp:1173]: just use cast<>

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (arsenm) [inline clang/lib/CodeGen/CGBuiltin.cpp:1191]: cast<> instead of isa+dyn_cast

Comment (Keenuts) [reply clang/lib/CodeGen/CGBuiltin.cpp:1191]: Oh right, thanks!

**Comment by Keenuts** [comment]
Hi, thanks for the reviews so far!
I believe we are ready to move forward on our side. Adding @llvm-beanz for the HLSL part.
The builtin I added is mostly to get something we can generate those intrinsics for. I am fine changing the name, or the implementation around it, it's just to have a wave an easy intrinsic to use/test with (no input, simplest one)

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (arsenm) [inline clang/lib/CodeGen/CGBuiltin.cpp:1136]: Don't see the point of wrapping this in optional instead of just returning null. It's no safer and now requires more syntax

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (arsenm) [inline clang/lib/CodeGen/CGBuiltin.cpp:1137]: It feels wrong that you would need to do a scan of a block to find this

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (arsenm) [inline clang/lib/CodeGen/CGBuiltin.cpp:1150]: Can you avoid bundleArgs and just construct the array inline?

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (arsenm) [inline clang/lib/CodeGen/CGBuiltin.cpp:1171]: Don't understand this. I thought it was a verifier error to be not-convergent and have a convergence token.

This won't actually do anything for an intrinsic. Because convergent is broken/backwards, you can't locally override it at a callsite and you can't remove the convergent from the underlying declaration

File: clang/lib/CodeGen/CGCall.cpp
Comment (arsenm) [inline clang/lib/CodeGen/CGCall.cpp:5690]: Can you drop the getCalledFunction part?

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (Keenuts) [reply clang/lib/CodeGen/CGBuiltin.cpp:1136]: No strong opinion. My rational is if you read "optional", you know it's not always there. But I'm fine with the pointer, changed to pointer.

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (Keenuts) [reply clang/lib/CodeGen/CGBuiltin.cpp:1137]: Yep, but I'm not sure I have any other option outside from some kind of caching(either attribute of a BB, or BB->I map). If it's this kind of alternative you are thinking about, maybe this should be left in a subsequent PR as an optimization?

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (Keenuts) [reply clang/lib/CodeGen/CGBuiltin.cpp:1150]: yep, thanks

Comment (Keenuts) [reply clang/lib/CodeGen/CGBuiltin.cpp:1150]: well, no... tried, and there is an overload which takes either an ArrayRef of a vector, so the inline array will cause an ambiguity :/

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (Keenuts) [reply clang/lib/CodeGen/CGBuiltin.cpp:1171]: good catch, I have no idea why this is here. I recall something around validation, but today reading it, it does not make sense. Thanks!

File: clang/lib/CodeGen/CGCall.cpp
Comment (Keenuts) [reply clang/lib/CodeGen/CGCall.cpp:5690]: Yes, thanks

File: clang/lib/Headers/hlsl/hlsl_intrinsics.h
Comment (llvm-beanz) [inline clang/lib/Headers/hlsl/hlsl_intrinsics.h:1303]: We should probably also apply `__attribute__((convergent))` to this function declaration. I know we didn't do that on the `WaveActiveCountBits` declaration, but we also didn't properly implement codegen for that.

File: clang/include/clang/Basic/Builtins.td
Comment (llvm-beanz) [inline clang/include/clang/Basic/Builtins.td:4560]: We probably want to add support for adding `Convergent` here.

Comment (Keenuts) [reply clang/include/clang/Basic/Builtins.td:4560]: Yes, but this implies touching to the mangling pattern..

File: clang/lib/Headers/hlsl/hlsl_intrinsics.h
Comment (Keenuts) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h:1305]: Yes, good catch, done

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (ssahasra) [reply clang/lib/CodeGen/CGBuiltin.cpp:1137]: This kind of map will be needed in more places as we continue to introduces uses of these convergence tokens. I have a simple Analysis in my upstreaming pipeline that produces such a map, but it's really very simple. You can wait for that, or just feel free to start your own.

Comment (Keenuts) [reply clang/lib/CodeGen/CGBuiltin.cpp:1137]: Good to know! I'd be in favor to implement this map once we have a merged use case.
-> This way, we can have an independent NFC PR which focuses on this optimization/refactoring by adding either an analysis or the bits in `llvm::BasicBlock` to quickly fetch those.

File: clang/include/clang/Basic/Builtins.td
Comment (arsenm) [reply clang/include/clang/Basic/Builtins.td:4560]: We should remove all the explicit convergent markers. We should go forward with https://reviews.llvm.org/D69498. Once you have convergence in the language at all, every function must be assumed convergent by default

File: clang/lib/Headers/hlsl/hlsl_intrinsics.h
Comment (arsenm) [inline clang/lib/Headers/hlsl/hlsl_intrinsics.h]: Should remove the convergent attribute (we never should have added this). The language settings need to set IR convergent on everything

Comment (Keenuts) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h]: @llvm-beanz FYI. Seems like I'll remove it then.

**Review by sudonatalie — COMMENTED** [review]
LGTM in terms of how this affects the SPIR-V backend, but if others who've reviewed here could +1 when ready, that would be good feedback for the other parts of the codebase that this touches before merging.

**Review by sudonatalie — APPROVED** [review]

File: clang/lib/Headers/hlsl/hlsl_intrinsics.h
Comment (llvm-beanz) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h]: I'm not sure I entirely agree with @arsenm here.

From the LLVM-IR and compiler implementation perspective it is helpful to be able to expect that convergent is the default expected behavior, but from the language perspective for HLSL it is preferable to imply that convergence is the outlier because in the common case convergence isn't required.

I really don't want to tie the implementation of HLSL to a 4.5 year old PR that isn't merged.

@bogner, do you have thoughts here?

Comment (arsenm) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h]: Convergence requires all transitive callers to also be convergent. No real code is going to do this. The existence of convergence operations means you have a convergent language and must optimize to known-not-convergent

Comment (Keenuts) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h]: Right, so in that case, I'll add the convergent attribute again, and later down the road, once we have the noconvergent-default in place, we'll be able to flip this back.
Thanks all for the context and explanations!

**Comment by Keenuts** [comment]
Rebases on main (almost, HEAD is slightly broken), and added back the convergence attribute.
The backend changes are ready for this intrinsic.

File: clang/lib/Headers/hlsl/hlsl_intrinsics.h
Comment (llvm-beanz) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h]: > Convergence requires all transitive callers to also be convergent. No real code is going to do this. The existence of convergence operations means you have a convergent language and must optimize to known-not-convergent

In languages like OpenCL and HLSL, this requirement is actually fine and we don't need it explicitly annotated. OpenCL and HLSL are effectively single-source kernel languages and all code gets inlined. So we can do robust attribute propagation.

Comment (llvm-beanz) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h]: I filed an issue to track enabling `-fconvergent-functions` for HLSL (#86506).

**Comment by Keenuts** [comment]
@arsenm would you be fine with those codegen changes as-is? Given that the convergent/no-convergent switch will be done later, depending on when the required IR change is merged?

**Comment by arsenm** [comment]
> @arsenm would you be fine with those codegen changes as-is? Given that the convergent/no-convergent switch will be done later, depending on when the required IR change is merged?

Yes

**Comment by Keenuts** [comment]
@ssahasra it is up to you then 😊

**Review by llvm-beanz — APPROVED** [review]
LGTM too.

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (ssahasra) [inline clang/lib/CodeGen/CGBuiltin.cpp:1148]: In this and other places, "ConvergenceControlToken" is a better name than "Attr" ... it's not really an attribute.

Comment (ssahasra) [reply clang/lib/CodeGen/CGBuiltin.cpp:1148]: Also function names need to start with lower case.

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (ssahasra) [inline clang/lib/CodeGen/CGBuiltin.cpp:1160]: Other places use "LoopToken". Ending with "Token" certainly feels more consistent.

**Review by ssahasra — COMMENTED** [review]
LGTM, with a few nits.

For the record, I did not have the bandwidth to comment on the translation from HLSL to SPIR-V. But given that the generated IR passes the verifier, this seems like a correct and very interesting first use of convergence control.

**Review by ssahasra — APPROVED** [review]

**Comment by Keenuts** [comment]
Thanks all the the reviews! We have 3 LGTMs and an ack from Arsenm, so I'm going to rebase on main, wait for the bots & tests, and if all is green, merge this.

**Comment by Keenuts** [comment]
Local tests for SPIR-V & DXIL pass

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (farzonl) [inline clang/lib/CodeGen/CGBuiltin.cpp:5890]: I missed this PR. Can you move this to `Value *CodeGenFunction::EmitHLSLBuiltinExpr(unsigned BuiltinID, const CallExpr *E)`

**Pull request merged**
